tutorials/dask.md

```diff
@@ -0,0 +1,147 @@
+
+The first part of this tutorial demonstrates how to use [`dask`](https://docs.dask.org/en/latest/) and [`dask.delayed`](https://docs.dask.org/en/latest/delayed.html) (or [`dask_image`](https://github.com/dask/dask-image)) to feed napari image data "[lazily](https://en.wikipedia.org/wiki/Lazy_evaluation)": that is, the specific image file corresponding to the requested timepoint/channel is only read from disk at the moment it is required (based on the current position of the dimension sliders in `napari`).  Additionally, we will see that *any* function that takes a filepath and returns a `numpy` array can be used to lazily read image data.  This can be useful if you have a proprietary format that is not immediately recognized by napari (but for which you have at least some way of reading into a `numpy` array)
+
+In some cases, data must be further processed prior to viewing, such as a deskewing step in stage-scanning light sheet microscopes.  Or perhaps you'd like to apply some basic image corrections or ratiometry prior to viewing.  The second part of this tutorial demonstrates the use of the [`dask.array.map_blocks`](https://docs.dask.org/en/latest/array-api.html#dask.array.map_blocks) function to describe an arbitrary sequence of functions in a declarative manner that will be performed *on demand* as you explore the data (i.e. move the sliders) in `napari`.
+
+## using dask.delayed to load images
+
+If you have a function that can take a filename and return a `numpy` array, for instance `skimage.io.imread`, you can create a "lazy" version of that function by calling `dask.delayed` on the function itself.  This *new* function, when handed a filename, will not actually read the file until explicitly asked with the `compute()` method:
+
+```python
+from skimage.io import imread
+from dask import delayed
+
+lazy_imread = delayed(imread)
+reader = lazy_imread('/path/to/file.tif')  # doesn't actually read the file
+array = reader.compute()  # *now* it reads.
+```
+
+We can create a [Dask array](https://docs.dask.org/en/latest/array.html) of delayed file-readers for *all* of the files in our multidimensional experiment using the `dask.array.from_delayed` function and a [`glob`](https://docs.python.org/3/library/glob.html) filename pattern:
+
+```python
+from skimage.io import imread
+from dask import delayed
+import dask.array as da
+from glob import glob
+
+filenames = sorted(glob("/path/to/experiment/*.tif"))
+# read the first file to get the shape and dtype
+sample = imread(filenames[0])
+
+lazy_imread = delayed(imread)  # lazy reader
+lazy_arrays = [lazy_imread(fn) for fn in filenames]
+dask_arrays = [
+    da.from_delayed(delayed_reader, shape=sample.shape, dtype=sample.dtype)
+    for delayed_reader in lazy_arrays
+]
+# Stack into one large dask.array
+stack = da.stack(dask_arrays, axis=0)
+stack.shape  # (nfiles, nz, ny, nx)
```

Comment:
It might be useful here to write the the stack can be reshaped? So that e.g. if it's `nfiles, ny, nx` it can be reshaped with da.reshape to `nt, nz, ny, nx`.

Reply:
glad you brought that up!  I'm actually struggling with this at the moment with my 2-channel sample data.  If I do
```python
stack.reshape(600,2,64,256,280)
```
the operation works fine, but it's actually not the appropriate reshape (dimensions are still mixed).  What I need to do is:
```python
stack.reshape(2,600,64,256,280)
```
but that hangs indefinitely in napari... and I haven't yet figured out why.  not sure if it's trying to read all the data yet.  let's figure that out before posting this.

Reply:
What exactly hangs? What about casting a slice to a NumPy array, does that hang?

Reply:
I haven't had time to check yet... i'll dig into it when I get a chance (hopefully later today) and let you know if anything is still mysterious.

Reply:
ok, it's not napari that's hanging... it's definitely dask (or possibly `dask-image`), but i still don't know where/why:

```python
from dask_image.imread import imread
stack = imread('/Volumes/extra850/Deskewed/*tif') 
stack.reshape(600, 2, 64, 256, 280)[0].compute()  # fine
stack.reshape(2, 600, 64, 256, 280)[0].compute()  # 😴
```
if I send `ctrl-c` while it's hung there, I get:
```
Traceback (most recent call last):
  File "...site-packages/IPython/core/interactiveshell.py", line 3326, in run_code
    exec(code_obj, self.user_global_ns, self.user_ns)
  File "<ipython-input-8-dde93889c3ad>", line 1, in <module>
    stack.reshape(2,600,64, 256, 280)[0].compute()
  File "...site-packages/dask/base.py", line 165, in compute
    (result,) = compute(self, traverse=False, **kwargs)
  File "...site-packages/dask/base.py", line 436, in compute
    results = schedule(dsk, keys, **kwargs)
  File "...site-packages/dask/threaded.py", line 81, in get
    **kwargs
  File "...site-packages/dask/local.py", line 475, in get_async
    key, res_info, failed = queue_get(queue)
  File "...site-packages/dask/local.py", line 133, in queue_get
    return q.get()
  File "python3.7/queue.py", line 170, in get
    self.not_empty.wait()
  File "python3.7/threading.py", line 296, in wait
    waiter.acquire()
KeyboardInterrupt
```

Reply:
a simpler example:
```python
from dask import delayed
import dask.array as da

shape = (10,10)

@delayed
def f(a,b): 
    print(f"{a}, {b}") 
    return np.zeros(shape)

arr = da.stack(
    [
        da.from_delayed(f(a, b), shape=shape, dtype="float64")
        for a in range(10)
        for b in range(2)
    ]
)
arr.shape  # (20, 10, 10)
arr.reshape(2,10,10,10)[0,0].compute()
# prints                                                                   
# 0, 0
# 0, 1
# 1, 0
# 2, 0
# 2, 1
# 3, 1
# 3, 0
# 1, 1
# 4, 1
# 4, 0

arr.reshape(10,2,10,10)[0,0].compute()                                                                               
# prints
# 0, 0
# 0, 1
```

I was wondering if this has to do with the limitations on `da.reshape` stated in the [docs](https://docs.dask.org/en/latest/array-api.html#dask.array.

>This is a parallelized version of the `np.reshape` function with the following limitations:
>1. It assumes that the array is stored in row-major order
>2. It only allows for reshapings that collapse or merge dimensions like
`(1, 2, 3, 4) -> (1, 6, 4)` or `(64,) -> (4, 4, 4)`

Reply:
Sounds like a Dask issue. Would try cut this example down into something that can be run with Dask without the data provided (perhaps random data would work?). It may also be able to cut this down in size to something that is just much slower by (time). That would be even more useful for developers to look at as there is something to profile and optimize. If you are able to get anything close to either of these two, would raise on the Dask issue tracker and see what people say. 🙂

Reply:
thanks @jakirkham!  I was starting to create an issue but realized it's pretty much the same thing as dask/dask/issues/3650, and as Matt suggested there, using .visualize() shows the problem (so I'm a little hesitant to post this there, since it's "known" behavior).  

```python
⬇️ arr[0].visualize()  # one f() call
```
![arr0vis](https://user-images.githubusercontent.com/1609449/69349107-37671280-0c45-11ea-9c26-377458d7f216.png)

```python
⬇️ arr.reshape(10, 2, *shape)[0, 0].visualize()  # reshaped.shape[1] f() calls
```
![download-1](https://user-images.githubusercontent.com/1609449/69357424-0f7eab80-0c53-11ea-8bdc-3fb6f738eb53.png)


```python
⬇️ arr.reshape(2, 10, *shape)[0, 0].visualize()  # reshaped.shape[1] f() calls
```
![download](https://user-images.githubusercontent.com/1609449/69349145-48178880-0c45-11ea-99ec-91776ddbd336.png)


@jakirkham posted in that thread [wondering whether rechunking would help](https://github.com/dask/dask/issues/3650#issuecomment-399510447)... but in this case it doesn't:
```python
rechunked = arr.reshape(2, 10, *shape).rechunk((1, 1, *shape))
rechunked
# dask.array<rechunk-merge, shape=(2, 10, 10, 10), dtype=float64, chunksize=(1, 1, 10, 10)>

rechunked[0, 0].compute(optimize_graph=True)
# still calls f() 10 times
```

the only workaround I can think of is to _manually_ distinguish between axes with the glob pattern... making multiple dask arrays and then stacking them, which is rather ugly

Reply:
Further analysis of the problem would be a welcome addition. It sounds to me like what you are saying will further the discussion in that Dask issue. Would encourage you to post these findings there.